script: Support native C/C++ shared object scripting

honggyukim opened the pull request:
This patch adds a native script support that can use a shared object as
a script plugin.

The basic interface is same as python and luajit, but the benefit of
this native script is that it doesn't add any overhead from the
scripting engines.  It just bypasses the execution flow into the
dynamically loaded shared object.

The target shared object has to be compiled from C/C++ programs that
provides the following functions.
```
  void uftrace_begin(void);
  void uftrace_entry(struct script_context *sc_ctx);
  void uftrace_exit(struct script_context *sc_ctx);
  void uftrace_end(void);
```
The 'struct script_context' is same as the struct used inside uftrace so
it's just passed without any manipulation.

The example C/C++ files are in sctipts directory and the shared objects
can be compiled as follows.
```
  $ make -C scripts/
  make: Entering directory '/home/honggyu/work/uftrace/scripts'
  gcc -fPIC -shared -o simple_c.so simple.c
  gcc -fPIC -shared -o simple_cpp.so simple.cpp
  make: Leaving directory '/home/honggyu/work/uftrace/scripts'
```
The basic usage is as follows.
```
  $ uftrace record -S scripts/simple_c.so --no-libcall t-abc
  program begins...
  entry : main()
  entry : a()
  entry : b()
  entry : c()
  exit  : c()
  exit  : b()
  exit  : a()
  exit  : main()
  program is finished
```
The output looks exactly same as python and luajit scripts.

Signed-off-by: Honggyu Kim <honggyu.kp@gmail.com>

namhyung commented:
Yes, I like it.  Actually I thought about this some time ago, but didn't find real use cases.  And we should consider argument and other info for proper usage.

mattyclarkson commented on scripts/simple.c:
```diff
@@ -0,0 +1,37 @@
+#include <stdio.h>
+#include <stdint.h>
+
+struct script_context {
```
Might be worth putting this into a header that can be easily included in a plugin?

Would we need any versioning of the interface?

honggyukim replied:
Yes, we can think about it. I've just put the struct as it looks very simple to copy and paste.

My concern is about having `struct list` for argument info because it adds a dependency on the internal list implementation. It's not always required so user can decide not to use and remove the dependency. I will think more about it. Thanks for the comment.

mattyclarkson replied:
> looks very simple to copy and paste.

Depends if we're ever going to change the structure. I guess as long as we put new members at the end we will be fine.

Depends if you think it would be useful for plugins to know the argument types? Can always follow up with that later.

In the future we may want to have a `uftrace/plugin.h` (or similar) that includes all the necessary headers for parsing the argument and this structure. For now, it's likely fine to ignore the arguments.

honggyukim replied:
Thanks. `uftrace/plugin.h` looks good. We have to decide how to provide the header to users. I also prototyped another feature that needs to provide a header to user, but it's also pending now at https://github.com/namhyung/uftrace/pull/636.

I will think more about how to provide those headers.

namhyung replied:
I agree that we should provide a header for API/ABI compatibility and also it's good to use a versioning.  Then we should consider installing/loading the header and plugin binaries as well.

honggyukim replied:
I've made a draft change at c2c22b8c regarding versioning and providing a header `uftrace-script.h`.

The C/C++ scripts can be written by including `uftrace-script.h` and be compiled. The version can be checked before using the struct. We can change the major version when the layout is change and the minor version can be changed when we added a new members at the end of struct.

It's just a draft version, but would like to hear how you think about this approach.

honggyukim commented:
I will also think about providing a way to disable the uftrace shmem record behaviour when we only need to use this native script interface.

namhyung commented:
Yes, it's reasonable when we run it during record and want to reduce the overhead.  It might be handled in `uftrace_begin` or using `--disable` option.

honggyukim commented:
> And we should consider argument and other info for proper usage.

I didn't use `struct script_info` yet in this implementation because it introduces another internal data structure `struct strv` so it will make the user to have more complicated header.

How about replace `struct strv cmds` to just `const char *cmds` as a concatenated string? I think we don't have to make it complicated unnecessarily and the users can parse the info themselves. I've applied this change at fe80d3f2.

honggyukim commented on utils/uftrace-script.h:
```diff
@@ -0,0 +1,54 @@
+	struct si_version {
+		uint8_t major;
+		uint8_t minor;
+	} ver;
```
I thought about handling the major version change doesn't allow the previously compiled shared objects cannot be used.  But minor version change allows that the shared objects can be used in a higher version of uftrace.

In summary
- If major version is different, then uftrace rejects loading the shared object.
- If uftrace minor version is higher than the shared object, then it can use the shared object without a problem.
- If uftrace minor version is lower than the shared object, then uftrace rejects loading the shared object.

But I also think that we can simply use a single integer version because the layout will not be changed often and we don't actually distribute the plugin shared object binaries.

honggyukim replied:
I've just made it to use a single uftrace version for both struct.

honggyukim commented on cmds/record.c:
```diff
@@ -323,7 +323,7 @@ static void setup_child_environ(struct opts *opts, int argc, char *argv[])
 
 		for (i = 0; i < argc; i++)
-			args = strjoin(args, argv[i], "\n");
+			args = strjoin(args, argv[i], " ");
```
Ah. That's why `\n' was used here. I will change it back to `\n'.

honggyukim commented on check-deps/Makefile.check:
```diff
@@ -35,6 +35,8 @@ ifneq ($(wildcard $(srcdir)/check-deps/have_libluajit),)
 endif
 
+COMMON_CFLAGS += -DHAVE_LIBNATIVE
+
```
Sure. I will remove this.

honggyukim replied:
It's now removed, but also we have to always enable script feature so added ca0c6e1e0229c1aa509b3e20cc71e54d87f37ffe.

honggyukim commented on utils/script-native.c:
```diff
@@ -0,0 +1,109 @@
+
+static int native_uftrace_begin(struct script_info *info)
+{
+	pr_dbg("%s()\n", __func__);
```
I've just used for debugging. I will remove this.

honggyukim commented on scripts/Makefile:
```diff
@@ -1,6 +1,6 @@
 script:
 	gcc -I.. -fPIC -shared -o simple_c.so simple.c
-	gcc -I.. -fPIC -shared -o simple_cpp.so simple.cpp
+	g++ -I.. -fPIC -shared -o simple_cpp.so simple.cpp
```
ack.

honggyukim commented on scripts/info.c:
```diff
@@ -1,29 +1,5 @@
-	struct list_head	*argspec;
-#endif
-};
+#include "utils/uftrace-script.h"
```
I will make `include` directory and put the headers that will be installed together. If you want to put the headers under `/usr/include/uftrace` directory, the name can be simpler like `/usr/include/uftrace/script.h`.

honggyukim commented on utils/uftrace-script.h:
```diff
@@ -0,0 +1,54 @@
+#else
+#include <cstdint>
+#include <cstdbool>
+#endif
```
Sure. I will use same headers for both C/C++.

honggyukim commented on utils/uftrace-script.h:
```diff
@@ -0,0 +1,54 @@
+	} ver;
+	char			*name;
+	char			*version;
+	bool			record;
```
Yes, that's also possible. I also thought that it's wasteful to use a bool type here.

honggyukim replied:
> Maybe we can have 2 integers as bitmasks for them. One indicates current status - only the record bit is available for now.

For now, I've just made the existing `record` variable can be disabled by the given native script. If it's disabled by script, then uftrace do not actually record the data.

> The other can control the behavior like disable, or skip arguments. I think we should keep original behavior unless script changes the bitmasks.

I don't see any use case for disabling argument record because we can already do this with uftrace command. I will just leave it as is, but can bump up the script version when it's needed.

honggyukim commented on utils/uftrace-script.h:
```diff
@@ -0,0 +1,54 @@
+	 * struct list_head internally uses 'new' as a variable name so
+	 * it's not possible to use in C++.
+	 */
+#endif
```
We may have to convert the `struct list_head` to this type. Could you explain why you used `char data[]` here?

Otherwise, how about modifying `utils/list.h` that can be compilable in C++, then distribute it together because uftrace also uses GPLv2 same as `list.h` anyway.

honggyukim replied:
Thanks for the suggestion. I see that if we introduce another way of handling arguments then we better use this in `script-python.c` and `script-luajit.c` as well without using the current `struct list_head`.

It may take more time to refine them all so I didn't update it for now.

honggyukim commented:
I've updated this PR again as follows.
* aabc21d3 uftrace: Do not use C++ keywords as variable names
* b1d3ff22 script: Change script_info.cmds from struct strv to char*
* 306b5fd9 script: Move script_info and script_context out to include/script.h
* edeb132a script: Support native C/C++ shared object scripting
* ab138715 script: Add version info to script_info
* cb05efc9 script: Make native script disable uftrace record

namhyung commented on utils/script-native.c:
```diff
@@ -0,0 +1,106 @@
+
+static int native_atfork_prepare(void)
+{
+	fflush(stdout);
```
It's already done in the `atfork_prepare_handler()`.

honggyukim replied:
Right. I removed it.

namhyung commented on utils/script-native.h:
```diff
@@ -0,0 +1,28 @@
+
+struct script_info;
+
+#ifdef HAVE_LIBNATIVE
```
This is not needed.

honggyukim replied:
It's removed now.

namhyung commented on include/script.h:
```diff
@@ -5,8 +5,22 @@
+int uftrace_script_version(void)
+{
+	return SCRIPT_VERSION;
+}
```
A function definition in the header?  I'm afraid that a plugin can consist of multiple files.

honggyukim replied:
Hmm.. Otherwise, the script writer has to implement this. Could you recommend if you have a better idea?

namhyung replied:
I think it's script writer's responsibility to check the version number.

namhyung commented on cmds/script.c:
```diff
@@ -133,6 +133,7 @@ int command_script(int argc, char *argv[], struct opts *opts)
 	struct uftrace_data handle;
 	struct uftrace_task_reader *task;
 	struct script_info info = {
+		.script_version = SCRIPT_VERSION,
```
API version?  I don't want to repeat "script" in the script_info..

honggyukim replied:
I've changed it to `api_version`.

namhyung commented on utils/script-native.c:
```diff
@@ -66,6 +68,13 @@ static int load_native_api_funcs(const char *so_pathname)
+	INIT_NATIVE_API_FUNC(uftrace_script_version);
+	if (SCRIPT_VERSION != __uftrace_script_version()) {
+		pr_warn("incompatible script versions (uftrace %d != module %d)\n",
+			SCRIPT_VERSION, __uftrace_script_version());
```
Well.. I think we should have backward compatibility so old version of script can still be run.

honggyukim replied:
Yes, it's right. But that's why I made the major and minor version to handle this before. If we want to keep the backward compatibility then we must not change the layout of struct, but we can only append the data to the struct in `include/script.h`.

namhyung commented on scripts/info.c:
```diff
@@ -9,6 +9,9 @@ void uftrace_begin(struct script_info *sc_info)
 	printf("%s\n", sc_info->cmds);
+
+	/* disable actual data record by uftrace */
+	sc_info->record = false;
```
As I said before, I'd like to have two bitmasks to control the behavior.

honggyukim replied:
I can make it to use two bitmasks, but I don't see if it's useful to control argument recording so left it for now. I can update it in the next update.

namhyung replied:
Let's do this in a separate PR.  I think it's better to focus on enabling the native script here.  Also other scripting like Python might use this feature too.

namhyung reviewed (commented):
Thanks for your work!

honggyukim commented:
I've just pushed only after rebase to make the "force-pushed" to be clearer in the next update. I will update this PR a bit later.

honggyukim commented on include/script.h:
```diff
@@ -0,0 +1,30 @@
+};
+
+/* context information passed to script */
+struct script_context {
```
It's also done.

honggyukim commented on Makefile:
```diff
@@ -187,6 +187,7 @@ LIBMCOUNT_UTILS_SRCS += $(srcdir)/utils/debug.c $(srcdir)/utils/regs.c
 LIBMCOUNT_UTILS_SRCS += $(srcdir)/utils/rbtree.c $(srcdir)/utils/filter.c
 LIBMCOUNT_UTILS_SRCS += $(srcdir)/utils/demangle.c $(srcdir)/utils/utils.c
 LIBMCOUNT_UTILS_SRCS += $(srcdir)/utils/script.c $(srcdir)/utils/script-python.c $(srcdir)/utils/script-luajit.c
+LIBMCOUNT_UTILS_SRCS += $(srcdir)/utils/script-native.c
```
done

honggyukim commented on scripts/info.c:
```diff
@@ -0,0 +1,24 @@
+#include <stdio.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include "include/script.h"
```
If so, maybe we should put `script.h` at `uftrace/include/uftrace/script.h`, otherwise we cannot compile this before installation.

namhyung replied:
I'm ok to put the header file there.  This script is an example that users can compile even without the uftrace source IMHO.  So it should work with the system-installed header.

honggyukim commented on scripts/simple.cpp:
```diff
@@ -0,0 +1,31 @@
+
+#if __cplusplus
+extern "C" {
+#endif
```
Yes, It's removed.

honggyukim commented on utils/script-native.c:
```diff
@@ -0,0 +1,106 @@
+#ifdef HAVE_LIBNATIVE
```
It's removed.

honggyukim commented:
@namhyung I've updated this patch again with the following changes.
- Split `uftrace_script_context` into `uftrace_script_base_ctx` and `uftrace_script_args`.
- Remove `list.h` dependency by passing `uftrace_script_base_ctx` to native script APIs without argument info.

In addition, I've already made comments for API version handling and record bits so please have a look again. I actually don't have a better idea yet so please let me know if you have better ideas.

namhyung commented on utils/script.h:
```diff
@@ -31,20 +31,29 @@ struct uftrace_script_info {
 
+/* context and args information passed to script */
+struct uftrace_script_context {
+	struct uftrace_script_base_ctx ctx;
```
I'd like to rename it to 'base' so that I can refer to 'ctx->base.tid' or so.

namhyung commented on Makefile:
```diff
@@ -338,6 +340,9 @@ endif
 	$(Q)$(INSTALL) $(objdir)/libmcount/libmcount-fast-single.so $(DESTDIR)$(libdir)/libmcount-fast-single.so
+	$(call QUIET_INSTALL, headers)
+	$(Q)$(INSTALL) -m 644 $(srcdir)/include/script.h $(DESTDIR)$(incdir)/script.h
+	$(Q)$(INSTALL) -m 644 $(srcdir)/utils/list.h $(DESTDIR)$(incdir)/list.h
```
We don't use it anymore, right?

namhyung commented on utils/script.h:
```diff
@@ -14,8 +14,6 @@
 #include "utils/utils.h"
 
-#define SCRIPT_ENABLED (SCRIPT_LUAJIT_ENABLED || SCRIPT_PYTHON_ENABLED)
-
```
Maybe we can just make this 'true', then no need to change other code.

namhyung commented on scripts/Makefile:
```diff
@@ -0,0 +1,7 @@
+script:
+	gcc -I.. -fPIC -shared -g -o simple_c.so simple.c
+	g++ -I.. -fPIC -shared -g -o simple_cpp.so simple.cpp
```
No need to implement same one twice.  Please add other one like 'count.so'?

namhyung commented on include/script.h:
```diff
@@ -0,0 +1,25 @@
+	char			*name;
+};
+
+#endif	/* UFTRACE_INCLUDE_SCRIPT_H */
```
Don't we need 'extern "C"' here?

namhyung commented on libmcount/mcount.c:
```diff
@@ -74,6 +74,9 @@ static int __maybe_unused mcount_depth = MCOUNT_DEFAULT_DEPTH;
 static bool __maybe_unused mcount_enabled = true;
 
+/* boolean flag to turn on/off recording, but keep script working */
+bool __maybe_unused record_enabled = true;
```
IIRC, mcount_enable can be used for this (hmm.. but it'd need this too)

```diff
diff --git a/libmcount/mcount.c b/libmcount/mcount.c
index a2310798..60b10945 100644
--- a/libmcount/mcount.c
+++ b/libmcount/mcount.c
@@ -1232,7 +1232,7 @@ void mcount_exit_filter_record(struct mcount_thread_data *mtdp,
                        mtdp->record_idx--;
 
                if (!mcount_enabled)
-                       return;
+                       goto out;
 
                if (!(rstack->flags & MCOUNT_FL_RETVAL))
                        retval = NULL;
@@ -1276,6 +1276,7 @@ void mcount_exit_filter_record(struct mcount_thread_data *mtdp,
                                mtdp->nr_events = k;  /* invalidate sync events */
                }
 
+out:
                /* script hooking for function exit */
                if (SCRIPT_ENABLED && script_str)
                        script_hook_exit(mtdp, rstack);
```

namhyung commented on utils/script.h:
```diff
@@ -32,20 +32,29 @@ struct uftrace_script_info {
+/* context and args information passed to script */
+struct uftrace_script_context {
+	struct uftrace_script_base_ctx base;
+	struct uftrace_script_args args;
```
I don't think this is needed.  Maybe we can add a separate script context definition for native script and leave this as is.

namhyung commented on utils/script.h:
```diff
@@ -29,7 +29,7 @@ struct uftrace_script_info {
 	char *version;
 	bool record;
-	struct strv cmds;
+	const char *cmds;
```
Let's not expose this to users.  Native scripts can have a separate header with your change.  Maybe argc + argv[] are more appropriate for native scripts.

namhyung commented on utils/script-luajit.c:
```diff
@@ -223,6 +223,9 @@ static int luajit_uftrace_begin(struct uftrace_script_info *info)
 {
 	int i;
 	char *s;
+	struct strv sv = {
```
Please use `STRV_INIT`.

namhyung commented on utils/script-python.c:
```diff
@@ -572,6 +572,9 @@ int python_uftrace_begin(struct uftrace_script_info *info)
 	char *s;
+	struct strv sv = {
+		0,
+	};
```
Ditto.

namhyung commented on uftrace.c:
```diff
@@ -1217,7 +1217,12 @@ void parse_script_opt(struct uftrace_opts *opts)
+		"",
+		"#",
+		"--",
+		"//",
```
I think the trailing comma makes line break.

namhyung commented on scripts/simple.cpp:
```diff
@@ -0,0 +1,26 @@
+{
+	std::cout << "program is finished\n";
+}
+}
```
This can be confusing.  Please add a comment like `// extern "C"`.

namhyung commented on scripts/Makefile:
```diff
@@ -0,0 +1,23 @@
+CSO     = $(SRCS:%.c=%_c.so)
+CPPSO   = $(SRCS:%.cpp=%_cpp.so)
+
+all: $(CSO) $(CPPSO)
```
This needs to be called from the top level Makefile.  But I think you should split the addition of examples to a separate commit.

namhyung reviewed (commented):
Thanks for working on this.

paranlee commented:
I thought it would be very useful to write a script to view CPU register information, and I'll work on that feature if it needs to be enhanced.

paranlee commented:
I've also rebased @honggyukim branch and test the rust plugin works well!
```
$ uftrace record -S  scripts/simple_rs.so --no-libcall tests/t-abc
uftrace_begin called scripts/simple_rs.so
uftrace_entry called with name: main
uftrace_entry called with name: a
uftrace_entry called with name: b
uftrace_entry called with name: c
uftrace_exit called with name: c
uftrace_exit called with name: b
uftrace_exit called with name: a
uftrace_exit called with name: main
uftrace_end called
```
- https://github.com/paranlee/uftrace/tree/review/native-script

```rust
use std::os::raw::c_char;
use std::ffi::CStr;

// Information passed during initialization
#[repr(C)] // Ensure C-compatible layout
pub struct UftraceScriptInfo {
	pub api_version: i32,
	pub name: *const c_char, // Use raw pointer for name and version
	pub version: *const c_char,
	pub record: bool,
	pub cmds: *const c_char,  // Use raw pointer for commands string
}

// Base context information passed to script
#[repr(C)] // Ensure C-compatible layout
pub struct UftraceScriptBaseCtx {
	pub tid: i32,
	pub depth: i32,
	pub timestamp: u64,
	pub duration: u64,
	pub address: usize, // Use usize for address on most platforms
	pub name: *const c_char, // Use raw pointer for name
}

#[no_mangle]
pub extern "C" fn uftrace_begin(info: &UftraceScriptInfo) {
	println!("uftrace_begin called {}", unsafe { CStr::from_ptr(info.name) }.to_string_lossy());
}

#[no_mangle]
pub extern "C" fn uftrace_entry(ctx: &UftraceScriptBaseCtx) {
	println!("uftrace_entry called with name: {}", unsafe { CStr::from_ptr(ctx.name) }.to_string_lossy());
}

#[no_mangle]
pub extern "C" fn uftrace_exit(ctx: &UftraceScriptBaseCtx) {
	println!("uftrace_exit called with name: {}", unsafe { CStr::from_ptr(ctx.name) }.to_string_lossy());
}

#[no_mangle]
pub extern "C" fn uftrace_end() {
	println!("uftrace_end called");
}
```

namhyung commented:
> I thought it would be very useful to write a script to view CPU register information

I'm not sure if it's useful.  The register content is volatile - IOW it's changed whenever you call a function (at least for some registers).

honggyukim commented:
> I've also rebased @honggyukim branch and test the rust plugin works well!

Thanks!

> I'll work on that feature if it needs to be enhanced.

I think current status is good but this PR is pending because we have many things to consider to keep the backward compatibility especially for argument handling. Maybe we better drop argument handling for now and let this merged without it.

namhyung commented:
> Maybe we better drop argument handling for now and let this merged without it.

Then it needs to be specified with API version.  I thought converting all data into a string and pass them once.  This would reduce the complexity to handle different types and number of arguments.  But the script should parse the argument (which may be fine since it can control the argument when recording).

yskelg commented:
This is especially great for uncertain post-mortem analysis or when we are unsure about which filtering parameters to narrow down.

In my opinion, the strength of this feature lies in the fact that uftrace currently cannot dynamically observe information at runtime, but with this, we can script specific parts as needed. Additionally, if we can implement such modules using Rust, C, or C++, which have lower performance overhead compared to other scripting languages, it allows for more flexible observation of the program's behavior at runtime. And once the source code is written as a module, I think it's great that anyone can share and reuse it.

@honggyukim's slides were a great inspiration for me. Thank you so much!
- https://uftrace.github.io/talks/uftrace_BHAS23_Arsenal.pdf
<img width="1249" alt="image" src="https://github.com/user-attachments/assets/5c76ffcc-3210-4112-844a-05d10bec539a">
<img width="1320" alt="image" src="https://github.com/user-attachments/assets/547b6f35-eb17-4212-91a4-4340860deb21">
<img width="1101" alt="image" src="https://github.com/user-attachments/assets/e37bd80a-12b1-45ac-9df3-9c7b1e76e28d">
<img width="1490" alt="image" src="https://github.com/user-attachments/assets/c70d9d47-b761-4c1e-9afa-775e103fa4c0">

I believe that people will find better use cases than the ones I come up with, and that's how it will be.

honggyukim commented:
Thanks @yskelg. I would like to complete this feature but I haven't had much time for working on this. It's in my TODO list for a long time but didn't get a strong request as of yet.

I can't promise I will continue in the near future but will come back to this when I can find some time. Thanks.
